Upmerge tfm 1 5

SebastianBoe opened the pull request:
Upmerge all three TF-M repo's from 1.4.1 to 1.5.0.

Also apply a fix for a breaking change.

SebastianBoe committed:
manifest: Update TF-M repos from 1.4.1 to 1.5.0

Update all three TF-M repos from 1.4.1 to 1.5.0.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

SebastianBoe commented:
DNM until further testing has been done

SebastianBoe committed:
tfm: kconfig: Fix config regression from TF-M 1.5

In TF-M 1.5 some defaults changed. Some of our CMake code was
unfortunately making assumptions about the defaults.

This is now rectified. The TF-M CMake variables that had their
defaults changed are now always set dependent on the Kconfig values
and we no longer leave the CMake variables unset and with default
values.

A cleanup will be done later to not rely on default values for all
variables. This patch only fixes the defaults that have changed.

Also, TFM_PSA_API is removed as it is now an internal variable.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

joerchan commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_manifest_list.yaml.in:
```diff
@@ -17,7 +17,6 @@
       "manifest": "${APPLICATION_SOURCE_DIR}/dummy_partition/tfm_dummy_partition.yaml",
       "output_path": "${TFM_BINARY_DIR}/dummy_partition",
       "tfm_partition_ipc": true,
-      "conditional": "TFM_PARTITION_DUMMY_PARTITION",
```
This commit is already merged to main, should rebase and get rid of it.

SebastianBoe replied:
done

SebastianBoe commented:
The remaining failures are all the mps2_an521 board failing a PS test. It times out, so it's not obvious what or where the failure is.

I don't have this board so I am unable to debug.

joerchan commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -32,8 +32,12 @@ set(TFM_CRYPTO_MODULES
 if (CONFIG_BUILD_WITH_TFM)
   if (CONFIG_TFM_IPC)
     list(APPEND TFM_CMAKE_ARGS -DTFM_PSA_API=ON)
```
When they changed this they made TFM_PSA_API an internal variable, so this should be removed I think, according to how they document their options.
https://github.com/zephyrproject-rtos/trusted-firmware-m/commit/23a583c3e4b62b13c51eecb0e84f4c312af51ab4#diff-683a66548742d1b9880d0f96ed9d3d425064064320616f67d977d3c917e4ebb6L19

theotherjimmy committed:
boards: mpsX-anY: Allow more time between watchdog resets in qemu

mps2_an521 and mps3_an547 need more time with TFM 1.5 to pass their
tests. This change gives it 2x as many instructions by changing the
tick speed of the watchdog to half it's prior frequency.

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>
Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

theotherjimmy commented:
The test failure seems to be caused by the watchdog resetting the test before it's done. Apply the following patch:
```
From 53cf1d6e8df496b76c74a5547be50660807fce2a Mon Sep 17 00:00:00 2001
From: Jimmy Brisson <jimmy.brisson@linaro.org>
Date: Thu, 16 Dec 2021 14:55:02 -0600
Subject: [PATCH] boards: mps2-an521: Allow more time between watchdog resets
 in qemu

This board needs more time with TFM 1.5 to pass it's tests. This
change gives it 2x as many instructions by changing the tick speed
of the watchdog to half it's prior frequency.

Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>
---
 boards/arm/mps2_an521/mps2_an521_ns_defconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/boards/arm/mps2_an521/mps2_an521_ns_defconfig b/boards/arm/mps2_an521/mps2_an521_ns_defconfig
index a9f603251b..8143c45480 100644
--- a/boards/arm/mps2_an521/mps2_an521_ns_defconfig
+++ b/boards/arm/mps2_an521/mps2_an521_ns_defconfig
@@ -12,7 +12,7 @@ CONFIG_CORTEX_M_SYSTICK=y
 CONFIG_RUNTIME_NMI=y
 CONFIG_TRUSTED_EXECUTION_NONSECURE=y
 CONFIG_ARM_MPU=y
-CONFIG_QEMU_ICOUNT_SHIFT=7
+CONFIG_QEMU_ICOUNT_SHIFT=6
 
 # GPIOs
 CONFIG_GPIO=y
-- 
2.33.1
```

My notes from debugging this can be found here: https://gist.github.com/theotherjimmy/f18b534fca1fdacaf6fb3279930f7ea4

SebastianBoe commented:
It would have been nice with a watchdog handler and a printf.

I don't know how mps2_an521+qemu does it, but on nrf you get an IRQ before the watchdog triggers so you can clean up important resources before the watchdog triggers.

SebastianBoe commented:
Applied the patch.

Great work @theotherjimmy .

microbuilder reviewed (changes requested):
Can you rebase this so that the `mps3_an547_ns` board is available for testing?

joerchan reviewed (changes requested):
-DTFM_PSA_API should be removed, it is no longer a cache variable, but an internal variable.

SebastianBoe commented:
@joerchan , @microbuilder : Done and done

SebastianBoe commented:
Applied @theotherjimmy 's fix to mps3_an547_ns as well as it showed up in CI as a failure.

Are there any more boards we should preventatively fix?

```
sebo@sebo:~/ncs/zephyr$ ack CONFIG_QEMU_ICOUNT_SHIFT=7
file.patch
24:-CONFIG_QEMU_ICOUNT_SHIFT=7

boards/arm/mps3_an547/mps3_an547_defconfig
14:CONFIG_QEMU_ICOUNT_SHIFT=7

boards/arm/mps2_an521/mps2_an521_remote_defconfig
14:CONFIG_QEMU_ICOUNT_SHIFT=7

boards/arm/mps2_an521/mps2_an521_defconfig
14:CONFIG_QEMU_ICOUNT_SHIFT=7

boards/arm/mps2_an385/mps2_an385_defconfig
12:CONFIG_QEMU_ICOUNT_SHIFT=7
sebo@sebo:~/ncs/zephyr$ cd boards/arm
sebo@sebo:~/ncs/zephyr/boards/arm$ ack CONFIG_QEMU_ICOUNT_SHIFT=
mps3_an547/mps3_an547_defconfig
14:CONFIG_QEMU_ICOUNT_SHIFT=7

mps3_an547/mps3_an547_ns_defconfig
15:CONFIG_QEMU_ICOUNT_SHIFT=6

mps2_an521/mps2_an521_remote_defconfig
14:CONFIG_QEMU_ICOUNT_SHIFT=7

mps2_an521/mps2_an521_ns_defconfig
15:CONFIG_QEMU_ICOUNT_SHIFT=6

mps2_an521/mps2_an521_defconfig
14:CONFIG_QEMU_ICOUNT_SHIFT=7

qemu_cortex_r5/qemu_cortex_r5_defconfig
4:CONFIG_QEMU_ICOUNT_SHIFT=3

qemu_cortex_m0/qemu_cortex_m0_defconfig
6:CONFIG_QEMU_ICOUNT_SHIFT=6

mps2_an385/mps2_an385_defconfig
12:CONFIG_QEMU_ICOUNT_SHIFT=7

qemu_cortex_m3/qemu_cortex_m3_defconfig
12:CONFIG_QEMU_ICOUNT_SHIFT=6

```

microbuilder commented:
@SebastianBoe I believe the AN521 and AN547 are the two MPSx boards we care about in this case. Thanks for the AN547 fix.

SebastianBoe commented:
CI shows that the same fix did not fix mps3_an547_ns unfortnuately. Timed out again.

SebastianBoe commented:
I've debugged the new failure somewhat.

During ITS configuration we wait for a TFM_INTERNAL_TRUSTED_STORAGE_SERVICE_SIGNAL signal, but we get a different signal and do a psa_panic. Which triggers a system reset, which triggers a usage fault somehow.

The error handling is so consistently bad ...

Failures should be loud, but TF-M consistently tries to persist through them. Which just causes the time of failure to move later in the execution.

NB: It is the second panic invocation that triggers, not the first in the below code.

```
psa_status_t tfm_its_req_mngr_init(void)
{
#ifdef TFM_PSA_API
    psa_signal_t signals = 0;

    if (tfm_its_init() != PSA_SUCCESS) {
        psa_panic();
    }

    while (1) {
        signals = psa_wait(PSA_WAIT_ANY, PSA_BLOCK);
        if (signals & TFM_INTERNAL_TRUSTED_STORAGE_SERVICE_SIGNAL) {
            its_signal_handle(TFM_INTERNAL_TRUSTED_STORAGE_SERVICE_SIGNAL);
        } else {
            psa_panic();
        }
    }
#else
    if (tfm_its_init() != PSA_SUCCESS) {
        return PSA_ERROR_GENERIC_ERROR;
    }
    its_is_init = true;
    return PSA_SUCCESS;
#endif
}

```

theotherjimmy commented:
I've reproduced the failures locally. Let me see if I can come up with any fixes.

SebastianBoe commented:
Hi,

when the non-secure FW tries to configure the FPU we get a
usagefault. Which is strange, because using FPU is only
disallowed when the SPE uses FPU[0], and the SPE is configured to
not use the FPU in this test.

Interestingly, the issue only occurs when IPC is enabled, in
library mode there is no problem with configuring FPSCR.

The same issue can be reproduced with sample hello_world on board
mps3_an547_ns with the following config:

CONFIG_BUILD_WITH_TFM=y
CONFIG_FPU=y
CONFIG_TFM_IPC=y

The FPU is otherwise configured in the same way in both the
failing and passing case.

I don't know why it would trigger a usage fault in one case, but
not the other.

[0] from 0x0106178e in __set_FPSCR+2 at /home/sebo/ncs/modules/hal/cmsis/CMSIS/Core/Include/cmsis_gcc.h:1622
[1] from 0x0106178e in z_arm_floating_point_init+42 at /home/sebo/ncs/zephyr/arch/arm/core/aarch32/prep_c.c:138
[2] from 0x0106178e in z_arm_prep_c+62 at /home/sebo/ncs/zephyr/arch/arm/core/aarch32/prep_c.c:182
[3] from 0x01061b72 in z_arm_reset at /home/sebo/ncs/zephyr/arch/arm/core/aarch32/cortex_m/reset.S:154

theotherjimmy commented:
Good timing! I just got to the same backtrace.

theotherjimmy committed:
Tests: disable mps3_an547_ns fpu sharing test

TFM disables the FPU on this target, so the test will usagefault
with a NOCP exception type when run.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>
Signed-off-by: Jimmy Brisson <jimmy.brisson@linaro.org>

theotherjimmy commented:
I recomend:
```diff
From b26f8c85319aecaac8ae3a636b44da753499f180 Mon Sep 17 00:00:00 2001
From: Jimmy Brisson <jimmy.brisson@linaro.org>
Date: Mon, 20 Dec 2021 09:57:39 -0600
Subject: [PATCH] Tests: disable mps3_an547_ns fpu sharing test

TFM disables the FPU on this target, so the test will usagefault
with a NOCP exception type when run.
---
 tests/arch/arm/arm_thread_swap_tz/testcase.yaml | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/tests/arch/arm/arm_thread_swap_tz/testcase.yaml b/tests/arch/arm/arm_thread_swap_tz/testcase.yaml
index a875735431..5466b02609 100644
--- a/tests/arch/arm/arm_thread_swap_tz/testcase.yaml
+++ b/tests/arch/arm/arm_thread_swap_tz/testcase.yaml
@@ -3,7 +3,11 @@ common:
   tags: arm fpu tfm
   arch_allow: arm
 tests:
-  arch.arm.swap.tz: {}
+  arch.arm.swap.tz:
+    # NOTE: this platform disables FPU access in TFM.
+    platform_exclude: mps3_an547_ns
   arch.arm.swap.tz_off:
     extra_configs:
       - CONFIG_ARM_NONSECURE_PREEMPTIBLE_SECURE_CALLS=n
+    # NOTE: this platform disables FPU access in TFM.
+    platform_exclude: mps3_an547_ns
-- 
2.33.1
```

SebastianBoe commented:
Updated PR with proposed patch

microbuilder commented:
@SebastianBoe I added a DNM tag since this has two approving reviews, but still has some issues, just as a precaution. Feel free to remove. Not sure why the ST board is failing in the post build script now, but the sign-off issue is easy to fix.

SebastianBoe commented:
@microbuilder : Thank you.

I have pushed a hotfix for the nucleo board.

It is most likely not the best solution, but unless someone can swiftly find a better solution I would prefer it if it was applied, and then re-worked together with the appropriate domain experts after the upmerge.

hotfix:
https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/64

SebastianBoe commented:
Green CI @microbuilder  :)

erwango committed:
samples/tfm_integration: nucleo_l552ze_q: Update flash partition

Due to a recent change in TFM, some more space should be allocated
to mcuboot flash partition (some space should be allocate for OTP)
(Cf commit db07170a34f ("Platform: Allocate space in flash for OTP")
in trusted-firmware-m repo)
Take this into account and increase mcuboot flash partition for
nucleo_l552ze_q_ns target.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

SebastianBoe committed:
tests: arm_thread_swap_tx: nucleo_l552ze_q: Update flash partition

Due to a recent change in TFM, some more space should be allocated
to mcuboot flash partition (some space should be allocated for OTP)
(Cf commit db07170a34f ("Platform: Allocate space in flash for OTP")
in trusted-firmware-m repo)
Take this into account and increase mcuboot flash partition for
nucleo_l552ze_q_ns target.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

microbuilder commented:
@SebastianBoe Grr `371/569 mps2_an521_ns             tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls FAILED Timeout (qemu)` ... didn't this pass before? Perhaps another watchdog issue where we need to slow execution down?

microbuilder commented:
> @SebastianBoe Grr `371/569 mps2_an521_ns tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls FAILED Timeout (qemu)` ... didn't this pass before? Perhaps another watchdog issue where we need to ~slow execution down~ speed QEMU execution up (as per the earlier AN547 fix)?

Hmm, the checks pass, but looking at the log I see a failure here:

```
 INFO    - /__w/zephyr/zephyr/twister-out/mps2_an521_ns/tests/kernel/mem_protect/syscalls/kernel.memory_protection.syscalls/handler.log
ERROR   - [INF] Beginning TF-M provisioning
 [WRN]  TFM_DUMMY_PROVISIONING is not suitable for production!  This device is NOT SECURE 
 [Sec Thread] Secure image initializing!
 Booting TFM v1.5.0
 Creating an empty ITS flash layout.
Creating an empty PS flash layout.
[Crypto] TF-M in library mode uses a dummy NV seed. This is not suitable for production! This device is NOT SECURE
*** Booting Zephyr OS build v2.7.99-2761-gdda988c846c0  ***
Running test suite syscalls
===================================================================
START - test_string_nlen
 PASS - test_string_nlen in 0.1 seconds
===================================================================
START - test_string_nlen
E: syscall z_vrfy_string_nlen failed check: Memory region 0x281008f0 (size 24) read access denied
 PASS - test_string_nlen in 0.1 seconds
===================================================================
START - test_to_copy
E: syscall user_copy failed check: Memory region 0x281008d0 (size 32) write access denied
 PASS - test_to_copy in 0.1 seconds
===================================================================
START - test_user_string_copy
E: string too long 0x10e1d7 (32)
E: syscall user_copy failed check: Memory region 0x281008f0 (size 24) read access denied
 PASS - test_user_string_copy in 0.2 seconds
===================================================================
START - test_user_string_alloc_copy
E: string too long 0x10e1d7 (32)
E: syscall z_user_alloc_from_copy failed check: Memory region 0x281008f0 (size 24) read access denied
 PASS - test_user_string_alloc_copy in 0.2 seconds
===================================================================
START - test_arg64
 PASS - test_arg64 in 0.1 seconds
===================================================================
START - test_more_args
 PASS - test_more_args in 0.1 seconds
===================================================================
START - test_syscall_torture
Running syscall torture test with 4 threads on 1 cpu(s)
```

It looks like the same timeout issues we had earlier, requiring the `CONFIG_QEMU_ICOUNT_SHIFT=6` change for AN521 and AN547 to avoid a watchdog reset. Since the results of the test suite aren't returned, it presumably doesn't find the FAILED string it's expecting, so the test passes.

I'm OK with merging this PR in, though, and resolving the AN521 timing issues separately since this is largely a CI problem, and TF-M 1.5.0 is working as expected, and I don't want to miss the merge window you have for tomorrow.

microbuilder commented:
@erwango Thanks for the time on this pulling you in just after the holidays.

I think this can be merged, and we'll sort out the final AN521 CI issue in a new PR since it isn't a TF-M issue, but I believe it's a QEMU timing problem and out of scope here.

SebastianBoe commented:
I agree.

Thank you for being accommodating.
